### PR TITLE
Branch extractor catalog wireframe viewlogs

### PR DESCRIPTION
**Pull request by bingzhang**
## Description

when user clicks view log link, Clowder will show the graylog log of this particular extractor to this user.

The test will need the graylog stack (elasticsearch, logspout and graylog) running. And proxy configuration of graylog in Clowder custom.conf.

There is a wiki page to have the script and command to setup Graylog stack and Clowder settings on the same machine: https://opensource.ncsa.illinois.edu/confluence/x/IgVTCg

I will continue to edit this wiki page.

## Review Time Estimate

- [ ] Immediately
- [ ] Within one week
- [x] When possible

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) 
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the CHANGELOG.md.
- [ ] I have signed the CLA
- [x] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by bingzhang** [comment]
working in progress. right now it shows extractor info. I will change to show graylog log.

**Comment by lmarini** [comment]
@bingzhang in the future you can create a draft PR (similar to adding the WIP in the title). I don't know if you can switch to one once it is created. https://github.blog/2019-02-14-introducing-draft-pull-requests/

**Comment by lmarini** [comment]
@bingzhang can you add a description of what changed and how we can test it?

**Comment by bingzhang** [comment]
> @bingzhang in the future you can create a draft PR (similar to adding the WIP in the title). I don't know if you can switch to one once it is created. https://github.blog/2019-02-14-introducing-draft-pull-requests/

Yes, It seems there is no option to switch to draft pull request. But this is a temporary pull request. I might delete it and make a new one.

File: conf/routes [hunk 1]
```diff
@@ -172,7 +172,7 @@ POST           /spaces/submit
 # EXTRACTORS
 # ----------------------------------------------------------------------
 GET            /extractors/:extractorName                                               @controllers.Extractors.showExtractorInfo(extractorName)
-
+GET            /extractorlog/:extractorName                                             @controllers.Extractors.showExtractorLog(extractorName)
```
Comment (lmarini) [inline conf/routes:175]: Can we change this to `/extractors:/extractorName/logs` to be more RESTful?

Comment (bingzhang) [reply conf/routes:175]: yes

**Comment by bingzhang** [comment]
Screen Shot 2020-09-14 at 10 13 30 AM

File: app/services/GraylogService.scala [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Service to get extraction logs from Graylog.
+ */
+class GraylogService @Inject() extends LogService {
+  def getLog(extractorName: String, submissionID: Option[String]): List[String] = {
+    val playConfig = play.Play.application().configuration()
+    val serviceEndpoint = playConfig.getString("clowder.log.serviceEndpoint")
+    val username = playConfig.getString("clowder.log.username")
+    val password = playConfig.getString("clowder.log.password")
+    val prefix = playConfig.getString("clowder.log.extractorNamePrefix")
```
Comment (bodom0015) [inline app/services/GraylogService.scala:31]: I would suggest adding some error-handling here.
If any of these are empty, we should probably return / throw an error about the missing configuration.

Right now, I get an obscure error message about a bad URL scheme if the `clowder.log.serviceEndpoint` value isn't set:
`The URI scheme, of the URI null/api/search/universal/relative?query=application_name%3Anullncsa.image.preview*, must be equal (ignoring case) to 'http', 'https', 'ws', or 'wss'`

Also notice the `null` string in the middle of the resulting URL if `clowder.log.extractorNamePrefix ` isn't set.

**Review by bodom0015 — COMMENTED** [review]
Seems like it will work, but I am unable to access the wiki page for the instructions on running Graylog and am therefore unable to test very thoroughly.

Do I need special permissions to view the wiki page?

File: app/services/GraylogService.scala (hunk 2 above)
Comment (bingzhang) [reply app/services/GraylogService.scala:31]: so, you define the variable as empty string or not defined in the conf file?

Comment (bingzhang) [reply app/services/GraylogService.scala:31]: do we allow clowder.log.extractorNamePrefix to be empty? or just check it is not defined.

**Comment by bingzhang** [comment]
> Seems like it will work, but I am unable to access the wiki page for the instructions on running Graylog and am therefore unable to test very thoroughly.
> 
> Do I need special permissions to view the wiki page?

Sorry, I think it has public access permission to the NCSA staff by default. I am not sure where to set it, but I will figure out.

**Comment by lmarini** [comment]
> Seems like it will work, but I am unable to access the wiki page for the instructions on running Graylog and am therefore unable to test very thoroughly.
> 
> Do I need special permissions to view the wiki page?

@bingzhang can you make the wiki page public and move it to the Clowder space?

**Comment by bodom0015** [comment]
Finally got Graylog up and running locally, but I'm still hitting some problems with the default configuration: `Execution exception in /Users/lambert8/workspace/syngenta/clowder/app/views/updateExtractors.scala.html:267`

In this case, my `custom.conf` did not have a definition for `viewextractorslogs`.  It would be helpful if the code or default config provided a value here instead of throwing an exception.

We should do one (or both) of the following:
* Change the code to properly handle when this key is empty (prevent the exception from being thrown)
* Apply a default of `false` to this key in `application.conf`

File: app/views/updateExtractors.scala.html [hunk 3]
```diff
@@ -264,11 +264,17 @@ <h1>Extractor Catalog</h1>
                                                 </li>
 
                                                 <!-- TODO: View Extractor Logs -->
+                                                @if(play.Play.application().configuration().getBoolean("viewextractorlogs")
+                                                        && play.Play.application().configuration().getString("clowder.log.serviceEndpoint")
+                                                        && play.Play.application().configuration().getString("clowder.log.username")
+                                                        && play.Play.application().configuration().getString("clowder.log.extractorNamePrefix")
```
Comment (bodom0015) [inline app/views/updateExtractors.scala.html:270]:
```suggestion
                                                @if(play.Play.application().configuration().getBoolean("viewextractorlogs", false)
                                                        && play.Play.application().configuration().getString("clowder.log.serviceEndpoint", "")
                                                        && play.Play.application().configuration().getString("clowder.log.username", "")
                                                        && play.Play.application().configuration().getString("clowder.log.extractorNamePrefix", "")
```

Comment (robkooper) [reply app/views/updateExtractors.scala.html:270]: why do we need the `viewextractorlogs` boolean. Will we ever define the other vars, and set vi viewextractorlogs to false?

**Comment by bingzhang** [comment]
> > Seems like it will work, but I am unable to access the wiki page for the instructions on running Graylog and am therefore unable to test very thoroughly.
> > Do I need special permissions to view the wiki page?
> 
> @bingzhang can you make the wiki page public and move it to the Clowder space?


> Finally got Graylog up and running locally, but I'm still hitting some problems with the default configuration: `Execution exception in /Users/lambert8/workspace/syngenta/clowder/app/views/updateExtractors.scala.html:267`
> 
> In this case, my `custom.conf` did not yet have a definition for `viewextractorslogs`. It would be helpful if the code or default config provided a value here instead of throwing an exception.
> 
> We should do one (or both) of the following:
> 
> * Change the code to properly handle when this key is empty (prevent the exception from being thrown)
> * Apply a default of `false` to this key in `application.conf`



> > Seems like it will work, but I am unable to access the wiki page for the instructions on running Graylog and am therefore unable to test very thoroughly.
> > Do I need special permissions to view the wiki page?
> 
> @bingzhang can you make the wiki page public and move it to the Clowder space?

I already made this public access, and Mike can follow the instruction to setup graylog stack. I just have conversion about tweak and polish the document.

**Review by bodom0015 — APPROVED** [review]

File: app/views/updateExtractors.scala.html (hunk 3 above)
Comment (lmarini) [reply app/views/updateExtractors.scala.html]: @bingzhang any thoughts regarding this?

Comment (bingzhang) [reply app/views/updateExtractors.scala.html]: fixed

**Pull request merged**